core/src/test/java/org/sonarsource/sonarlint/core/container/storage/FileMatcherTest.java

```diff
@@ -103,6 +100,46 @@ public void should_return_most_common_prefixes() {
     assertThat(match.sqPrefix()).isEqualTo(Paths.get("sq1"));
   }
 
+  @Test
+  public void should_favor_deepest_common_path() {
+    List<Path> idePaths = Arrays.asList(
+      Paths.get("local1/pom.xml"),
+      Paths.get("local1/build.properties"),
+      Paths.get("local1/src/main/java/com/foo/A.java"));
+
+    List<Path> sqPaths = Arrays.asList(
+      Paths.get("sq1/pom.xml"),
+      Paths.get("sq1/build.properties"),
+      Paths.get("sq2/src/main/java/com/foo/A.java")
+
+    );
+    FileMatcher.Result match = fileMatcher.match(sqPaths, idePaths);
+    assertThat(match.idePrefix()).isEqualTo(Paths.get("local1"));
+    assertThat(match.sqPrefix()).isEqualTo(Paths.get("sq2"));
+  }
+
+  @Test
+  public void should_disfavor_path_having_multiple_matches() {
+    List<Path> idePaths = Arrays.asList(
+      Paths.get("local1/pom.xml"),
+      Paths.get("local1/build.properties"),
+      Paths.get("local1/src/A.java"));
+
+    List<Path> sqPaths = Arrays.asList(
+      Paths.get("sq1/pom.xml"),
+      Paths.get("sq1/build.properties"),
+      Paths.get("sq2/pom.xml"),
+      Paths.get("sq2/build.properties"),
+      Paths.get("sq3/pom.xml"),
+      Paths.get("sq3/build.properties"),
+      Paths.get("sq4/src/A.java")
+
+    );
+    FileMatcher.Result match = fileMatcher.match(sqPaths, idePaths);
+    assertThat(match.idePrefix()).isEqualTo(Paths.get("local1"));
+    assertThat(match.sqPrefix()).isEqualTo(Paths.get("sq4"));
+  }
+
```

Comment:
```suggestion

  @Test
  public void should_match_first_similar_hierarchy() throws Exception {
    List<Path> idePaths = Arrays.asList(
      Paths.get("local/sub/index.html"),
      Paths.get("local/sub/product1/index.html"),
      Paths.get("local/sub/product2/index.html"),
      Paths.get("local/sub/product3/index.html")
    );

    List<Path> sqPaths = Arrays.asList(
      Paths.get("sq/index.html"),
      Paths.get("sq/news/index.html"),
      Paths.get("sq/news/product1/index.html"),
      Paths.get("sq/news/product2/index.html"),
      Paths.get("sq/news/product3/index.html"),
      Paths.get("sq/products/index.html"),
      Paths.get("sq/products/product1/index.html"),
      Paths.get("sq/products/product2/index.html"),
      Paths.get("sq/products/product3/index.html"),
      Paths.get("sq/company/index.html"),
      Paths.get("sq/company/jobs/index.html"),
      Paths.get("sq/company/news/index.html"),
      Paths.get("sq/company/contact/index.html")
    );

    FileMatcher.Result match = fileMatcher.match(sqPaths, idePaths);
    assertThat(match.idePrefix()).isEqualTo(Paths.get("local/sub"));
    assertThat(match.sqPrefix()).isEqualTo(Paths.get("sq/news"));
  }
```

I ran this on my machine to check the stability of the behavior in ambiguous cases, I think it might be relevant to add it for documentation purposes.

Reply:
As discussed I don't think it is testing more cases than other existing tests, but I will add it as a "more realistic" test to have a safety net.